### ISSUE TITLE
Rule id 930110 Bypass

### ISSUE THREAD

**Issue by donky16** [coreruleset/coreruleset#2140]
### Description
Because of the conversion of cmdLine function, ..\ will not be intercepted.
poc
![image](https://user-images.githubusercontent.com/24637333/123362812-d67f8900-d5a3-11eb-858b-506a39faabd1.png)

### Environment

* CRS version: v3.4/dev
* Paranoia level: 1
* Web Server and version: apache 2.4.29
* Operating System and version: Ubuntu 18.04

**Comment by dune73** [comment]
Hey @donky16, thank you for reporting. I confirm this is not caught in our default installation (in fact it's only caught at PL4). 

However, I am not sure this can be considered an attack. If I set a to `..\etc\passwd` we detect it just fine. So it boils down to the question whether CRS should block access to for a parameter with a value `secret` that might point to a file - or not.

**Comment by donky16** [comment]
..\..\secret will be caught, but ..\secret not. I think this is a contradiction. And whatever the secret is, ..\secret should be considered as a path travel attack. That's why rule 930110 exists.

**Comment by dune73** [comment]
Thank you. We looked too much at `secret` and not enough at `..\`. So it's in fact `..\<anything>` that should be detected as attack. That is a worthwhile complaint.

**Comment by lifeforms** [comment]
Good catch. Probably the failure to trigger on this pattern is due to `t:cmdLine` which deletes the backslash. We should get an additional test for the false negative, and then see if the removal of the transformation passes the test.

**Comment by RedXanadu** [comment]
@lifeforms I've spent some time looking at this. Unfortunately, I don't think the answer is as simple as removing the `t:cmdLine` transformation function (I tried that, without success.)

When I submit the following test request:
`test.com/?arg=..\testing`
I see in the debug log:
`[9] Target value: "/?arg=..\\testing"`

Are backslashes escaped internally _before_ any operators are executed?

Based on how `\` was being turned into `\\`, as a test, I tweaked the regular expression of rule 930110. I replaced all instances of
`[\\/]`
with the following:
`(?:\\\\|/)`
to make it look for `\\` or `/`.

The modified rule successfully picked up the test request:
`test.com/?arg=..\testing`

So, this amended regular expression seems to work as intended:
```
"@rx (?:(?:^|(?:\\\\|/))\.\.(?:\\\\|/)|(?:\\\\|/)\.\.(?:(?:\\\\|/)|$))"
```

Existing regular expression for comparison:
```
"@rx (?:(?:^|[\\/])\.\.[\\/]|[\\/]\.\.(?:[\\/]|$))"
```

**Comment by dune73** [comment]
@lifeforms : Do you agree with @RedXanadu's analysis. Time for a PR?

**Comment by fzipi** [comment]
@RedXanadu Will this work for nginx and modsec3, considering https://github.com/SpiderLabs/ModSecurity/issues/2148?

**Comment by RedXanadu** [comment]
@fzipi Ah-hah! I think you may have solved it. Thanks for sharing that issue: I thought I had seen something about Apache escaping posted somewhere, but couldn't find it again :)

Changing the regular expression from its current form:
`(?:^|[\\/])\.\.(?:[\\/]|$)`
to this:
`(?:^|[\\\\/])\.\.(?:[\\\\/]|$)`
works!

Based on the information from https://github.com/SpiderLabs/ModSecurity/issues/2148, I think that when _Apache_ interprets the current regular expression for 930110, it is turned from `[\\/]` into `[\/]`, which is then interpreted simply as `[/]`.

I wonder if 930110 was originally written and tested with nginx/libmodsecurity? Maybe it works there (I can't test that at the moment.)

**Comment by fzipi** [comment]
Excellent. If we can have a summary written, I can create a couple of negative/positive tests and run those through our testing process and see if this works for both.

**Comment by RedXanadu** [comment]
## Summary

This false negative for rule 930110 only affects Apache: the rule works correctly with nginx. The issue appears to be related to the different ways that Apache and nginx require "\\" (backslash) characters to be represented (as explained in detail [here](https://github.com/SpiderLabs/ModSecurity/issues/2148)).

### Test 1: Using rule 930110 with its current regular expression

The current regular expression:
`@rx (?:^|[\\/])\.\.(?:[\\/]|$)`

Test request:
`> GET /?test=..\pineapple HTTP/1.1`

Response from **Apache**:
`< HTTP/1.1 200 OK` ❌ False negative

Response from **nginx**:
`< HTTP/1.1 403 Forbidden` ✔

```
2021/08/18 18:29:20
    [info] 26#26: *4 ModSecurity: Warning. Matched "Operator `Rx' with parameter `(?:^|[\\/])\.\.(?:[\\/]|$)' against variable `ARGS:test' (Value: `..\pineapple' )
    [file "/etc/modsecurity.d/owasp-crs/rules/REQUEST-930-APPLICATION-ATTACK-LFI.conf"]
    [line "52"]
    [id "930110"]
    [rev ""]
    [msg ""]
    [data ""]
    [severity "0"]
    [ver "OWASP_CRS/3.3.2"]
    [maturity "0"]
    [accuracy "0"]
    [hostname "172.21.0.4"]
    [uri "/"]
    [unique_id "1629311360"]
    [ref "o0,3v11,12"], client: 172.21.0.1, server: localhost, request: "GET /?test=..\pineapple HTTP/1.1", host: "localhost:81"
```

### Test 2: Using rule 930110 with a modified regular expression

Regular expression that uses a representation of "\\" (backslash) that's portable between Apache and nginx/libmodsecurity:
`@rx (?:^|[\\\\/])\.\.(?:[\\\\/]|$)`

Test request:
`> GET /?test=..\pineapple HTTP/1.1`

Response from **Apache**:
`< HTTP/1.1 403 Forbidden` ✔

```
[Wed Aug 18 19:00:22.148086 2021]
    [:error]
    [pid 16:tid 139804062770944]
    [client 172.26.0.1:39018]
    [client 172.26.0.1] ModSecurity: Warning. Pattern match "(?:^|[\\\\\\\\/])\\\\.\\\\.(?:[\\\\\\\\/]|$)" at ARGS:test.
    [file "/etc/modsecurity.d/owasp-crs/rules/REQUEST-930-APPLICATION-ATTACK-LFI.conf"]
    [line "71"]
    [id "930110"]
    [msg "Path Traversal Attack (/../)"]
    [data "Matched Data: ..\\x5c found within ARGS:test: ..\\x5cpineapple"]
    [severity "CRITICAL"]
    [ver "OWASP_CRS/3.3.2"]
    [tag "modsecurity"]
    [tag "application-multi"]
    [tag "language-multi"]
    [tag "platform-multi"]
    [tag "attack-lfi"]
    [tag "paranoia-level/1"]
    [tag "OWASP_CRS"]
    [tag "capec/1000/255/153/126"]
    [hostname "localhost"]
    [uri "/"]
    [unique_id "YR1YxrU65M48UPt4PGB0AQAAAIE"]
```

Response from **nginx**:
`< HTTP/1.1 403 Forbidden` ✔

```
2021/08/18 19:00:01
    [info] 26#26: *1 ModSecurity: Warning. Matched "Operator `Rx' with parameter `(?:^|[\\\\/])\.\.(?:[\\\\/]|$)' against variable `ARGS:test' (Value: `..\pineapple' )
    [file "/etc/modsecurity.d/owasp-crs/rules/REQUEST-930-APPLICATION-ATTACK-LFI.conf"]
    [line "52"]
    [id "930110"]
    [rev ""]
    [msg ""]
    [data ""]
    [severity "0"]
    [ver "OWASP_CRS/3.3.2"]
    [maturity "0"]
    [accuracy "0"]
    [hostname "172.26.0.4"]
    [uri "/"]
    [unique_id "1629313201"]
    [ref "o0,3v11,12"], client: 172.26.0.1, server: localhost, request: "GET /?test=..\pineapple HTTP/1.1", host: "localhost:81"
```

**Comment by RedXanadu** [comment]
The regular expression for rule 930110 is actually slightly different in v3.4/dev (it's been split into two halves).

Re-running the above tests to be safe yielded the same results. The same logic still applies: changing `\\` to `\\\\`.

I'll submit a PR for consideration shortly (to amend the regular expression)

**Comment by fzipi** [comment]
Once the PR is there, I can help with writing tests for this one.

**Comment by dune73** [comment]
I'm losing track a bit here, but AFAICS, we don't have a PR for this, so far?

Please correct me if I'm wrong.

**Comment by RedXanadu** [comment]
@dune73 I think this specific false negative has been fixed, with a new test also added, in #2178

I believe this issue can be closed now

**Comment by dune73** [comment]
Thank you. I suspected as much and overlooked your link to this issue in the PR message for #2178. Closing now.
